core/templates/variant_allocator.h

```diff
@@ -0,0 +1,294 @@
+/**************************************************************************/
+/*  variant_allocator.h                                                   */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef VARIANT_ALLOCATOR_H
+#define VARIANT_ALLOCATOR_H
+
+#include "core/core_globals.h"
+#include "core/os/spin_lock.h"
+#include "core/typedefs.h"
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdlib>
+#include <type_traits>
+
+#if defined(__GNUC__) || defined(__clang__)
+#if (__has_builtin(__builtin_popcountll))
+#define builtin_popcountll
+#endif
+#if (__has_builtin(__builtin_ffsll))
+#define builtin_ffsll
```

Comment:
Because we don't know for certain if `uint64_t` is `unsigned long long`, I'd rather use the generic equivalents:
```suggestion
#if __has_builtin(__builtin_popcountg)
#define builtin_popcountg
#endif
#if __has_builtin(__builtin_ffsg)
#define builtin_ffsg
```